[DataPipe] Add `.set_length` method to IterDataPipe

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #83750

Fixes https://github.com/pytorch/data/issues/549

This allows users to manually set length of an `IterDataPipe` with no other side effect. This is useful for DataPipes whose final length cannot be known in advance (e.g. `filter`). If you know the final length with certainty, you can manually set it for usages by DataLoader or other DataPipes.

Previously, users theoretically could use `.header(length)` to manually set length. However, the UX is suboptimal in that 1) the name is not obvious, and 2) warnings are raised unless the `HeaderDataPipe` has been fully traversed once to confirm the length.

Differential Revision: [D38879334](https://our.internmc.facebook.com/intern/diff/D38879334)

NivekT committed:
[DataPipe] Add .set_length method to IterDataPipe

[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -365,6 +365,19 @@ def conditional_reset(*args, **kwargs):
 
             namespace['reset'] = conditional_reset
 
+        if '__len__' in namespace:
```
We need to add this hook no matter `__len__` is in the namespace or not. We can raise `NotImplementedError` if `_length` is None.

NivekT replied:
Done. Note that `isinstance(dp, collections.abc.Sized)` will always be `True` now, because `__len__` always exists as a method through the hook. Is that fine?

If that is not fine, we can consider somehow intercepting the `len(dp)` call but it may be complicated (if feasible at all).

NivekT replied:
Note that this causes so test failures, so we will need to adjust the expected test results if we stick with this implementation.

ejguan replied:
> If that is not fine, we can consider somehow intercepting the `len(dp)` call but it may be complicated (if feasible at all).

Based on my understanding, not really feasible as `__len__` is implemented on the class rather than instsances.

> Note that this causes so test failures, so we will need to adjust some other logic, such as the following in "combining.py": `if all(isinstance(dp, Sized) for dp in self.datapipes)`

Emmm, this is a good question. I personally don't know if there is any drawback of making all `IterDataPipe` `Sized` (raising `TypeError` if needed). Let's discuss with Vitaly about the decision? This seems like a BC-breaking change.

NivekT replied:
Yea, let's have @VitalyFedyunin add his thoughts. 

The benefit on implementing `.set_length` as a DataPipe instead is that it doesn't require changes to `__len__` or `Sized`.

NivekT committed:
Update on "[DataPipe] Add .set_length method to IterDataPipe"


This allows users to manually set length of an `IterDataPipe` with no other side effect. This is useful for DataPipes whose final length cannot be known in advance (e.g. `filter`). If you know the final length with certainty, you can manually set it for usages by DataLoader or other DataPipes.

Previously, users theoretically could use `.header(length)` to manually set length. However, the UX is suboptimal in that 1) the name is not obvious, and 2) warnings are raised unless the `HeaderDataPipe` has been fully traversed once to confirm the length.


[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Add .set_length method to IterDataPipe"


This allows users to manually set length of an `IterDataPipe` with no other side effect. This is useful for DataPipes whose final length cannot be known in advance (e.g. `filter`). If you know the final length with certainty, you can manually set it for usages by DataLoader or other DataPipes.

Previously, users theoretically could use `.header(length)` to manually set length. However, the UX is suboptimal in that 1) the name is not obvious, and 2) warnings are raised unless the `HeaderDataPipe` has been fully traversed once to confirm the length.


[ghstack-poisoned]

NivekT commented on torch/utils/data/datapipes/_typing.py:
```diff
@@ -365,6 +365,23 @@ def conditional_reset(*args, **kwargs):
+                if len_func is not None:
+                    return len_func(*args, **kwargs)
+                else:
+                    raise TypeError(f"object of type '{datapipe}' has no len()")
```
The original exception is a `TypeError` but I can switch it to `NotImplementedError` if that is preferred.

ejguan replied:
`TypeError` is needed for running `list(dp)`.

NivekT committed:
Update on "[DataPipe] Add .set_length method to IterDataPipe"


This allows users to manually set length of an `IterDataPipe` with no other side effect. This is useful for DataPipes whose final length cannot be known in advance (e.g. `filter`). If you know the final length with certainty, you can manually set it for usages by DataLoader or other DataPipes.

Previously, users theoretically could use `.header(length)` to manually set length. However, the UX is suboptimal in that 1) the name is not obvious, and 2) warnings are raised unless the `HeaderDataPipe` has been fully traversed once to confirm the length.


[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38879334).

NivekT commented:
We will be landing https://github.com/pytorch/data/pull/747 instead of this PR.
